core/chains/evm/logpoller/log_poller.go

```diff
@@ -63,6 +63,72 @@ type LogPoller interface {
 	LogsDataWordBetween(ctx context.Context, eventSig common.Hash, address common.Address, wordIndexMin, wordIndexMax int, wordValue common.Hash, confs Confirmations) ([]Log, error)
 }
 
+// GetLogsBatchElem hides away all the interface casting, so the fields can be accessed more easily, and with type safety
+type GetLogsBatchElem rpc.BatchElem
+
+func NewGetLogsReq(filter Filter) *GetLogsBatchElem {
+	topics := make2DTopics(filter.EventSigs, filter.Topic2, filter.Topic3, filter.Topic4)
+
+	params := map[string]interface{}{
+		"address": []common.Address(filter.Addresses),
+		"topics":  topics,
+	}
+
+	return &GetLogsBatchElem{
+		Method: "eth_getLogs",
+		Args:   []interface{}{params},
+		Result: new([]types.Log),
+	}
+}
+
+func (e GetLogsBatchElem) params() map[string]interface{} {
+	return e.Args[0].(map[string]interface{})
+}
+
+func (e GetLogsBatchElem) Addresses() []common.Address {
+	return e.params()["address"].([]common.Address)
+}
+
+func (e GetLogsBatchElem) SetAddresses(addresses []common.Address) {
```

Comment:
I'm not a fan of setter methods in general, cause it's very easy to make a mistake in concurrent env with mutable objects. What do you think about making it immutable (and thread-safe) by always constructing a new object? Right now, you can accidentally fall into case in which one routine reads from that struct and another one tries to update addresses. Errors like this are extremely difficult to spot 

You can achieve that by having an additional constructor that takes existing `GetLogsBatchElem` and `addresses []common.Address`

Reply:
I don't follow what you're saying about immutability or concurrency. These objects are only ever accessed by a single thread, so we shouldn't have to worry about either of those issues.

I tried to implement this originally with only having the getter methods, but it resulted in a lot of extra awkward casting. The setter is just to clean up that mess, so that it can be set in a simple way.

Reply:
I see... I think it could make sense because in this case, you'd have to go out of your way with some casting to directly access the fields so it wouldn't be easy to mess things up, despite not technically being immutable.  On the other hand, I think this is only used in parts of the code which are protected by a mutex.

One of my primary concerns with this PR was to avoid causing performance issues due to much more complex structures having to get torn down and rebuilt. So I tried to err on the side of not rebuilding anything that wasn't essential, in favor of updating existing structures. Probably doesn't matter too much in this case though because it's just a slice--I'll look it over again, thanks for the suggestion.